src/libcore/iter.rs

```diff
@@ -3270,6 +3270,50 @@ impl<A, B> DoubleEndedIterator for Zip<A, B> where
 ///
 /// [`map()`]: trait.Iterator.html#method.map
 /// [`Iterator`]: trait.Iterator.html
+///
+/// # Notes about side effects
+///
+/// The [`map()`] iterator implements [`DoubleEndedIterator`], meaning that
+/// you can also [`map()`] backwards:
+///
+/// ```rust
+/// let v: Vec<i32> = vec![1, 2, 3].into_iter().rev().map(|x| x + 1).collect();
+///
+/// assert_eq!(v, [4, 3, 2]);
+/// ```
+///
+/// [`DoubleEndedIterator`]: trait.DoubleEndedIterator.html
+///
+/// However, if you call `.rev()` to iterate in the other direction, it
+/// will only be 'backwards' if your closure has no state. In other words,
+/// something like this might act in a way that you wouldn't expect:
```

Comment:
I still feel like this text leaves too much to the reader. Some context: I think it's great to add a note here, because this can be really confusing. It definitely confounded me for a bit the first time I ran into it. In your example below, you never walk through what's happening, just say that it's the same, so the reader is left to step through the code on their own.

I would prefer an [example like this](http://is.gd/R3u5Xy). I included some comments that also give the explanatory notes I wanted:

```rust
fn main() {
    // Prints ('a', 1), ('b', 2), ('c', 3), as expected.
    let mut c = 0;
    for pair in vec!['a', 'b', 'c'].into_iter()
                                   .map(|letter| { c += 1; (letter, c) }) {
        println!("{:?}", pair);
    }

    // Now we add a call to `rev`. Prints ('c', 1), ('b', 2), ('a', 3).
    // Note that the letters are reversed, but the values of the counter
    // still go in order. This is because `map` is still being called lazilly
    // on each item, it's just that we are popping items off the back of
    // the vector now, instead of shifting them from the front.
    c = 0;
    for pair in vec!['a', 'b', 'c'].into_iter()
                                   .map(|letter| { c += 1; (letter, c) })
                                   .rev() {
        println!("{:?}", pair);
    }
}
```

(Also, isn't your code wrong? They don't, in fact, produce the same thing?)